cloudinit/config/cc_mounts.py

```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size):
+    """Size is in MB."""
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    def create_swap(fname, size, method):
+        errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+
+        LOG.debug("Creating swapfile in '%s' on fstype '%s' using '%s'",
+                  fname, fstype, method)
+
+        if method == "fallocate":
+            cmd = ['fallocate', '-l', '%dM' % size, fname]
+        elif method == "dd":
+            cmd = ['dd', 'if=/dev/zero', 'of=%s' % fname, 'bs=1M',
+                   'count=%d' % size]
+
+        try:
+            util.subp(cmd, capture=True)
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+
+            if method == "fallocate":
+                LOG.warning("Will attempt with dd.")
+                create_swap(fname, size, "dd")
+            else:
+                unlink_f(fname)
+
+    tdir = os.path.dirname(fname)
+    util.ensure_dir(tdir)
+
+    fstype = get_fstype(tdir)
+
+    if fstype in ("xfs", "btrfs"):
+        create_swap(fname, size, "dd")
+    else:
```

Comment:
perhaps before defaulting to fallocate, we should check that it exists?

Reply:
So as-written, `create_swap(..., "fallocate")` will fallback to using dd if using fallocate fails (which it obviously would if it's absent).

My suggestion in https://github.com/canonical/cloud-init/pull/70#discussion_r359064993 would make that clearer, I think.

Reply:
Done that.